### PR TITLE
Adds function which calculates sites selection frequency for a set of provided samples

### PR DESCRIPTION
**Pull request by Rosejoycrocker**
- allows plotting of site selection frequency choropleth maps using `create_map!` to plot frequencies.
- can take output of `find_robust()` to plot site selection frequency of robust scenarios.

### CODE REVIEW

**Comment by ConnectedSystems** [comment]
Hi Rose, are there commits missing in this PR? I only see a function that collates frequency of seeding deployments - no `create_map()` or similar.

**Comment by ConnectedSystems** [comment]
Marking this issue as contributing to resolving #254

**Comment by Rosejoycrocker** [comment]
Hi Takuya, I realised you can use your create_map!() function with a suitable geojson and data = frequencies calculated from seeded_sites_frequency() to plot the frequncy maps (as far as I can see).

**Comment by ConnectedSystems** [comment]
Oh I see, thanks!

File: src/analysis/site_selection.jl
Comment (ConnectedSystems) [reply src/analysis/site_selection.jl:19]: On `scens` being a `namedtuple`, don't make any changes now but I am flagging that this is probably a roundabout way of handling multiple cases.

I've used this pattern before, but I think in cases like this where there's only a single "class" of data (i.e., data for each RCP), it's probably better to make things more flexible:

```julia
function seeded_sites_frequency(rs::ResultSet, rcps...)
    # ids will be each entry in `rcps`
end
```

**Review by ConnectedSystems — CHANGES_REQUESTED** [review]
Requesting some minor adjustments. Some relate to making the user experience a little smoother

File: src/analysis/site_selection.jl
Comment (ConnectedSystems) [reply src/analysis/site_selection.jl:15]: Just realised - wouldn't this be better with reefs/sites as rows and RCPs as columns?

Not sure if the depiction above is accurate now either, it makes it look like an array of paired arrays.

File: src/analysis/site_selection.jl
Comment (Rosejoycrocker) [reply src/analysis/site_selection.jl:19]: I made it a NamedTuple to work with the output of find_robust, which I imagine will be the most common use case of plotting robust scenario site selection frequencies

Comment (ConnectedSystems) [reply src/analysis/site_selection.jl:19]: Yeah, I wasn't sure if that was the best way to do it  when I wrote that, was a bit rushed. Need to think a bit more about it.

File: src/analysis/site_selection.jl
Comment (Rosejoycrocker) [reply src/analysis/site_selection.jl:24]: I can't import n_locations?

Comment (ConnectedSystems) [reply src/analysis/site_selection.jl:24]: It doesn't exist for ResultSets yet, but it should still pull in the one for `Domain`. Just checking: where are you importing from?

Comment (Rosejoycrocker) [reply src/analysis/site_selection.jl:24]: oh right, I can write one for ResultSet

File: src/analysis/site_selection.jl
Comment (Rosejoycrocker) [reply src/analysis/site_selection.jl]: All good, should I change it to a variable number of inputs anyway?

Comment (ConnectedSystems) [reply src/analysis/site_selection.jl]: No, hold off for now - I/we need to discuss what it should look like and how it might ease/complicate other analyses

Comment (Rosejoycrocker) [reply src/analysis/site_selection.jl]: ok cool :)

Comment (Rosejoycrocker) [reply src/analysis/site_selection.jl]: how's intervention_frequency() for the function name and scen_indices for the the scenario index NamedTuple?

Comment (Rosejoycrocker) [reply src/analysis/site_selection.jl]: I could add a 'log' input which indicates shade of seed log so it does both.

Comment (ConnectedSystems) [reply src/analysis/site_selection.jl]: I like `intervention_frequency()`

```julia

function intervention_frequency(rs, scen_indices, log_type)
    # This is just an example, really I'd pull out ResultSet fieldnames that end in "_log"
    occursin(log_type, [:seed, :shade, :fog]) || DomainError("Unsupported log")
   
    log = getfield(rs, Symbol("$(log_type)_log")))

    # Extract data from `log`
    # ...
end
```

Comment (ConnectedSystems) [reply src/analysis/site_selection.jl]: `DomainError` is probably not the right error type to throw either - AttributeError? ValueError? I'll leave it to you.

Comment (Rosejoycrocker) [reply src/analysis/site_selection.jl]: cool, I'll add this, thanks

File: src/analysis/site_selection.jl
Comment (Rosejoycrocker) [reply src/analysis/site_selection.jl]: I put the `n_locations(rs)` function in the ResultSet file, but should it be with the Domain one?

**Comment by Rosejoycrocker** [comment]
Hey @ConnectedSystems, I think I've addressed the comments. I'm not sure I've used NamedDimsArray in the way you'd like for the output, so let me know if I should change it.

**Comment by Rosejoycrocker** [comment]
Ok updated the doc string and added axis keys, let me know if there's anything else

File: src/analysis/site_selection.jl
Comment (ConnectedSystems) [inline src/analysis/site_selection.jl:48]: Rose, please use the simpler method to creating named axes:

```julia
NamedDimsArray(zeros(n_locs, length(rcps)), locations=1:n_locs, rcps=rcps)
```

I'm not sure if this handles some edge case, so please double check that it's needed.
`[rcps[k] for k in eachindex(rcps)]`

File: src/analysis/site_selection.jl
Comment (ConnectedSystems) [inline src/analysis/site_selection.jl:52]: Is there an issue with using the dimension names?

**Review by ConnectedSystems — CHANGES_REQUESTED** [review]
Thanks Rose, one, possibly two, more changes

File: src/analysis/site_selection.jl
Comment (Rosejoycrocker) [reply src/analysis/site_selection.jl:48]: it errored previously without the vector creation because rcps is a Tuple

Comment (ConnectedSystems) [reply src/analysis/site_selection.jl:48]: Okay, but if it's a tuple of values isn't looping over each index a convoluted way of converting to a vector?

`[rcps[k] for k in eachindex(rcps)]`

vs

```
# Indicating to Julia what the type is makes a tiny bit faster
Symbol[rcps...]
```

Comment (ConnectedSystems) [reply src/analysis/site_selection.jl:48]: Or you could just make `rcps` a vector on L 44, then no conversion necessary

Comment (Rosejoycrocker) [reply src/analysis/site_selection.jl:48]: Just realised I can use `collect()`, I always forget you can use it on tuples of strings/symbols as well

File: src/analysis/site_selection.jl
Comment (ConnectedSystems) [inline src/analysis/site_selection.jl:37]: Wondering how this is being used on your end? I hit an undefined error, and I can't see this file being included anywhere - doesn't seem to be part of the ADRIA package?

Comment (Rosejoycrocker) [reply src/analysis/site_selection.jl:37]: sorry,I just realised I was supposed to ask you about this in the PR and forgot to- should the import of site_selection.jl be in the ADRIA.jl file or the analysis module or both?

**Comment by ConnectedSystems** [comment]
Cleaned things up for you


```julia
using ADRIA, Statistics, Plots

rs = ADRIA.load_results("some result set")

tac = ADRIA.metrics.scenario_total_cover(rs)
rsv = ADRIA.metrics.scenario_rsv(rs)

# Create matrix of mean scenario outcomes
mean_tac = vec(mean(tac, dims=1))
mean_sv = vec(mean(rsv, dims=1))
y = hcat(mean_tac, mean_sv)

# Find all pareto optimal scenarios where all metrics >= 0.9
rule_func = x -> all(x .>= 0.9)
robust_scens = ADRIA.analysis.find_robust(rs, y, rule_func, [45, 60])

# Retrieve seeding intervention frequency for robust scenarios
robust_selection_frequencies = ADRIA.analysis.intervention_frequency(rs, robust_scens, :seed)

bar(robust_selection_frequencies, labels=["RCP45" "RCP60"], xlabel="Sites", ylabel="Times Selected")
```

image

**Review by ConnectedSystems — APPROVED** [review]

File: src/analysis/site_selection.jl
Comment (Rosejoycrocker) [reply src/analysis/site_selection.jl]: Thanks, so analysis files only need to be imported in the analysis module?

Comment (ConnectedSystems) [reply src/analysis/site_selection.jl]: Yes - functions/methods should be grouped/compartmentalised by modules according to their purpose. Their files are grouped by folders where appropriate.

You should see this if you look at the package directory structure. You should be able to intuit what lives where, and how it all relates to each other. For example, all code for analysis live together, ditto for file handling, etc.

At the top level, we only need to import the sub-modules (and later, package extensions).
Then the user only needs to import ADRIA or the sub-modules as necessary.

Otherwise you get into a situation where it's a messy free-for-all.

**Pull request merged**
